### ISSUE TITLE
THREE.ColladaLoader: Undefined sampler

### ISSUE THREAD

**Issue by MohmadAyman** [mrdoob/three.js#12530]
##### Description of the problem 
Hi,
Sorry for ticketing so many issues lately :sweat_smile: 

Some meshes started showing this error after upgrading to the new Collada, e.g. [this mesh](https://bitbucket.org/osrf/gazebo_models/raw/ca22037469dcc6acdb568de11e322bfadf256d0e/turtlebot/meshes/68-02421-8000-RA_Turtlebot_F-F_Standoff.dae).

So after some debugging, in the loader [at this line](https://github.com/mrdoob/three.js/blob/dev/examples/js/loaders/ColladaLoader.js#L1389) it is trying to find `texture0` in the samplers, while the correct sampler is  `texture0-sampler` .

This workaround solves it for me.

```
var sampler = effect.profile.samplers[ textureObject.id ] || effect.profile.samplers[ textureObject.id + '-sampler' ];
```
 

##### Three.js version

- [x] Dev
- [ ] r87
- [ ] ...

##### Browser

- [x] All of them
- [ ] Chrome
- [ ] Firefox
- [ ] Internet Explorer

##### OS

- [x] All of them
- [ ] Windows
- [ ] macOS
- [ ] Linux
- [ ] Android
- [ ] iOS

**Comment by Mugen87** [comment]
I'm confused why the usage of the sample just fails in your case 🤔 . We tested with so many examples where this was no problem.

Your proposed fix might work in your case, but it does not look like a generic solution to me.  I need to debug this one...

**Comment by MohmadAyman** [comment]
After having a look at the rest of our models, it seems like there are only a very few `mesh`es that have this problem ... but was working fine with the old collada :confused:  .

**Comment by Mugen87** [comment]
Is it possible to provide the entire model as a zip file again? 😊 Makes it easier to see if the texture is actually applied.

**Comment by KavehR** [comment]
Having the same issue with Maya 2017 here. Is the loader expecting a specific shading network to be set before exportation with the DAE_FBX exporter from Maya ? I am asking because it seems, among a couple of secondary disparities, that the sampler2D newparam is not created in the dae file, nor passed as the texture in the texture tag of the desired channel. The texture is fed with the image id directly instead. 
Correcting the issue manually in the file solves the problem, but are we supposed to export a specific material network or is the collada exporter supposed to properly convert various Maya materials directly suitable to threejs's collada loader ?

**Comment by MohmadAyman** [comment]
Sure, here is the [model zipped](https://drive.google.com/file/d/0B60Ua3vn3uQ-ZkE5cXJ4eldBQkk/view?usp=sharing) .

**Comment by KavehR** [comment]
In your files, it looks like the blender exporter is properly giving the sampler's id to the channel's texture, whereas meshlab's present kind of the same problem as Maya 2017's here, giving the image id instead of the sampler id to the channel's texture. Feels like a collada exportation plugin problem..

**Comment by Mugen87** [comment]
Sketchfab for example is able to load the file correctly. I wonder how you can implement this feature without making assumptions about the name of the id. Just adding '-sampler' does not look very robust to me.

Unfortunately, i did not find something useful in the spec. It only refers to the workflow via samplers.

https://www.khronos.org/files/collada_spec_1_5.pdf

**Comment by cnspaha** [comment]
Run into the same problem with every object I downloaded from clara.io which uses textures.

BUT! Found following ;)
https://github.com/KhronosGroup/OpenCOLLADA/issues/92#issuecomment-13135708

Also in the collada spec 1.5 ( @Mugen87  )
texture="myParam" => The value is specified by a reference to a previously defined <sampler2D> object.
![image](https://user-images.githubusercontent.com/22448974/32658016-46e5e40e-c618-11e7-87b5-b20e17a13c6f.png)

So the implementation of the Collada Loader seems fine according to the specification... 
=> the issue itself is invalid since it rather seems like a bug in the exporter

Models from clara.io, with textures, works fine when uploaded again on clara.io so I don't know what clara.io does in perticular but probably they check for the sampler and if it is now available they create one... hacky but seems to work on their site.

**Comment by Mugen87** [comment]
@bhouston Can you recommend an approach for handle this issue inside `ColladaLoader`?

tl;dr: The problem is that some Collada files have `texture` tags that refer directly to an image and not to a sampler. `ColladaLoader` is currently unable to load such files because it expects a reference to a sampler.

**Comment by Mugen87** [comment]
I figured out a clean way that should solve the issue.

**Comment by MohmadAyman** [comment]
I can help trying it out when ready 👍

**Comment by bhouston** [comment]
BTW just to be clear, because I missed this originally, we use the Autodesk FBX SDK export to create the Collada files exported from Clara.io.  It has proven to be buggy.

**Comment by russdanner** [comment]
This model still produces the issue:
https://clara.io/view/20ed671f-f62c-4c54-b1c9-18758f48cbdf#

**Comment by Mugen87** [comment]
@russdanner Can you please shader the model?

**Comment by looeee** [comment]
@russdanner your model is loading fine for me. @Mugen87 here's the file
[heart-dae.zip](https://github.com/mrdoob/three.js/files/1824893/heart-dae.zip)

**Comment by Mugen87** [comment]
@looeee Thanks! Yeah, the editor loads the file without errors.

**Comment by LuckXP** [comment]
I'm having trouble with this model as well: [https://clara.io/view/1a03ac6b-d6b5-4c2d-9f1a-c80068311396#](https://clara.io/view/1a03ac6b-d6b5-4c2d-9f1a-c80068311396#)

**Comment by Mugen87** [comment]
@LuckXP What do you mean with "trouble"? Can you please specify what's going wrong?

**Comment by LuckXP** [comment]
@Mugen87 Oh. Yeah sorry. Just that the model I posted results in the same "Undefined sampler." console warning.

**Comment by Mugen87** [comment]
@LuckXP Can you please share the model in this thread?
